src/libraries/Microsoft.Extensions.Diagnostics/src/Metrics/ListenerSubscription.cs

```diff
@@ -0,0 +1,324 @@
+                {
+                    RefreshInstrument(instrumentPair.Key, instrumentPair.Value);
+                }
+            }
+        }
+
+        // Called under _instrument lock
+        private void RefreshInstrument(Instrument instrument, InstrumentStatus status)
+        {
+            var alreadyEnabled = status.Enabled;
+            var enable = false;
+            var rule = GetMostSpecificRule(instrument);
+            if (rule != null)
+            {
+                enable = rule.Enable;
+            }
+
+            if (!enable && alreadyEnabled)
+            {
+                status.Enabled = false;
+                _meterListener.DisableMeasurementEvents(instrument);
+            }
+            else if (enable && !alreadyEnabled)
+            {
+                // The first time we enable an instrument, we need to call InstrumentPublished.
+                if (!status.Published)
+                {
+                    // However, a listener might decline to enable the instrument, remember that.
+                    status.Published = true;
+                    status.ListenerEnabled = _metricsListener.InstrumentPublished(instrument, out var state);
+                    status.State = state;
+                }
+
+                if (status.ListenerEnabled)
+                {
+                    _meterListener.EnableMeasurementEvents(instrument, status.State);
+                    status.Enabled = true;
+                }
+            }
+        }
+
+        private InstrumentRule? GetMostSpecificRule(Instrument instrument)
+        {
+            InstrumentRule? best = null;
+            foreach (var rule in _rules)
+            {
+                if (RuleMatches(rule, instrument, _metricsListener.Name)
+                    && IsMoreSpecific(rule, best, isLocalScope: instrument.Meter.Scope != null))
+                {
+                    best = rule;
+                }
+            }
+
+            return best;
+        }
+
+        // internal for testing
+        internal static bool RuleMatches(InstrumentRule rule, Instrument instrument, string listenerName)
+        {
+            // Exact match or empty
+            if (!string.IsNullOrEmpty(rule.ListenerName)
+                && !string.Equals(rule.ListenerName, listenerName, StringComparison.OrdinalIgnoreCase))
+            {
+                return false;
+            }
+
+            // Exact match or empty
+            if (!string.IsNullOrEmpty(rule.InstrumentName)
+                && !string.Equals(rule.InstrumentName, instrument.Name, StringComparison.OrdinalIgnoreCase))
+            {
+                return false;
+            }
+
+            if (!(rule.Scopes.HasFlag(MeterScope.Global) && instrument.Meter.Scope == null)
+                && !(rule.Scopes.HasFlag(MeterScope.Local) && instrument.Meter.Scope != null)) // TODO: What should we be comparing Scope to, the DefaultMeterFactory / IMeterFactory?
```

Comment:
nit: is this can be simplified as following?

```C#
        if (!(instrument.Meter.Scope is null ? rule.Scopes.HasFlag(MeterScope.Global) : rule.Scopes.HasFlag(MeterScope.Local)))
```

Reply:
This has changed a bit so that local specifically checks if the scope matches the current meter factory.

@noahfalk, should a global rule match instruments from other rule factories, or just meters with no scope?

Reply:
My intent is that:
- Global => meter.Scope == null.
- Local => meter.Scope == IMeterFactory from current DI container

If the scope is anything other than those two values then it would never be a match.